### PR TITLE
[WIP] replace (most) log contexts with tracing spans

### PR DESCRIPTION
**Pull request by hawkw**
This branch replaces a majority of the proxy's _sui generis_ logging contexts with `tracing` spans. 

I probably need to rebase this against the latest master, so I'm opening it as a draft for now.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -101,6 +101,7 @@ trust-dns-resolver = { git = "https://github.com/bluejekyll/trust-dns", rev = "7
 tracing            = "0.1.9"
 tracing-futures    = "0.0.1-alpha.1"
 tracing-log        = "0.1"
+tracing-tower      = { git = "https://github.com/tokio-rs/tracing", branch = "eliza/tower-tweaks-two" }
```
Comment (olix0r) [inline Cargo.toml]: Planning to release this before merge?

Comment (hawkw) [reply Cargo.toml]: Hopefully; that crate is still somewhat experimental (a significant amount of it was written in support of this PR).

File: lib/linkerd2-router/src/lib.rs [hunk 2]
```diff
@@ -2,27 +2,30 @@
...
+use tracing_futures::{Instrumented, Instrument};
 
 mod cache;
 pub mod error;
 
 use self::cache::{Cache, PurgeCache};
 
 /// Routes requests based on a configurable `Key`.
-pub struct Router<Req, Rec, Mk>
+pub struct Router<Req, Rec, Mk, T = ()>
```
Comment (olix0r) [inline lib/linkerd2-router/src/lib.rs]: Does this really need to be generic?

**Comment by hawkw** [comment]
Obsoleted by #374

**Pull request closed without merging**